### PR TITLE
ISPN-5821 Add manual cluster switch in Java client

### PR DESCRIPTION
**Pull request by galderz**
https://issues.jboss.org/browse/ISPN-5821

* Java Hot Rod client can manually switch between pre-defined clusters
  via `switchToCluster` and `switchToDefaultCluster` methods.

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/tcp/TcpTransportFactory.java [hunk 1]
```diff
@@ -304,49 +305,67 @@ public void invalidateTransport(SocketAddress serverAddress, Transport transport
...
-         //1. first add new servers. For servers that went down, the returned transport will fail for now
-         for (SocketAddress server : addedServers) {
-            log.newServerAdded(server);
-            try {
-               connectionPool.addObject(server);
-            } catch (Exception e) {
-               if (!quiet) log.failedAddingNewServer(server, e);
-            }
-         }
+   // To be called from within `lock` synchronized block
```
Comment (danberindei) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/tcp/TcpTransportFactory.java]: Could use `net.jcip.annotations.GuardedBy` here.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/TransportFactory.java [hunk 2]
```diff
@@ -67,4 +67,6 @@
    void reset(byte[] cacheName);
 
    boolean trySwitchCluster(byte[] cacheName);
+
+   boolean trySwitchCluster(String clusterName);
```
Comment (danberindei) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/TransportFactory.java]: I think you need a new name, there's already a `trySwitchCluster(byte[] cacheName)`, which I presume handles the automatic switch.

**Comment by galderz** [comment]
@danberindei Fixed those comments

**Comment by galderz** [comment]
@pruivo Rebased

**Comment by danberindei** [comment]
Integrated, thanks Galder!

**Pull request closed without merging**